daisy_workflows/image_test/linux_common/utils.py

```diff
@@ -99,10 +172,204 @@ def RunTest(test_func):
 
 
 def SetupLogging():
+  """Configure Loggin system."""
   logging_level = logging.DEBUG
   logging.basicConfig(level=logging_level)
   console = logging.StreamHandler()
   console.setLevel(logging_level)
   logging.getLogger().addHandler(console)
 
+
 SetupLogging()
+
+
+class MetadataManager:
+  """Utilities to manage metadata."""
+
+  SSH_KEYS = 'ssh-keys'
+  SSHKEYS_LEGACY = 'sshKeys'
+  INSTANCE_LEVEL = 1
+  PROJECT_LEVEL = 2
+
+  def __init__(self, compute, instance, ssh_user='tester'):
+    """Constructor.
+
+    Args:
+      compute: object, from GetCompute.
+      instance: string, the instance to manage the metadata.
+      user: string, the user to create ssh keys and perform ssh tests.
+    """
+    self.zone = self.FetchMetadataDefault('zone')
+    self.project = self.FetchMetadataDefault('project')
+    self.compute = compute
+    self.instance = instance
+    self.md_obj = None
+    self.level = None
+    self.last_fingerprint = None
+    self.ssh_user = ssh_user
+
+  def FetchMetadata(self, level):
+    """Fetch metadata from the server. This also fetches the fingerprint
+    required for StoreMetadata.
+
+    Args:
+      level: enum, INSTANCE_LEVEL or PROJECT_LEVEL to fetch the metadata.
+    """
+    self.level = level
+    if level == self.PROJECT_LEVEL:
+      request = self.compute.projects().get(project=self.project)
+      md_id = 'commonInstanceMetadata'
+    else:
+      request = self.compute.instances().get(
+          project=self.project, zone=self.zone, instance=self.instance)
+      md_id = 'metadata'
+
+    @RetryOnFailure
+    def _FetchMetadata():
+      response = request.execute()
+      self.md_obj = response[md_id]
+      if self.md_obj['fingerprint'] != self.last_fingerprint:
+        self.last_fingerprint = self.md_obj['fingerprint']
+      else:
+        raise ValueError('FetchMetadata retrieved same fingerprint')
+    _FetchMetadata()
+
+  def StoreMetadata(self):
+    """Store Metadata previously fetched with FetchMetadata."""
+    if self.level == self.PROJECT_LEVEL:
+      request = self.compute.projects().setCommonInstanceMetadata(
+          project=self.project, body=self.md_obj)
+    else:
+      request = self.compute.instances().setMetadata(
+          project=self.project, zone=self.zone, instance=self.instance,
+          body=self.md_obj)
+    request.execute()
+    self.md_obj = None
+    self.level = None
+
+  def ExtractKeyItem(self, md_key):
+    """Extract a given key value from the metadata previously fetched
+    with FetchMetadata.
+
+    Args:
+      md_key: string, the key of the metadata value to be searched.
+
+    Returns:
+      md_item: dict, in the format {'key', md_key, 'value', md_value}.
+      None: if md_key was not found.
+    """
+    try:
+      md_item = (
+          md for md in self.md_obj['items'] if md['key'] == md_key).next()
```

Comment:
I don't understand this function. Can you load the metadata response as json and simplify a lot of this file?